Implement a rule command.

Unia opened the pull request:
I saw the TODO in `velox.c` and thought I'd have a go.

As you can see, I implemented a new `rule` command that can receive several types just like the `action` command. For now, I implemented only a `tag` type, which will move a window to a certain tag. Other types could for example be to automatically make windows float.

Two things have not been worked out yet. I wanted your feedback on this before making a decision myself:
* The current implementation does not support window titles spanning several words, e.g. "Mozilla Firefox". This is due to the matching being done with `strtok_r` separating on whitespaces. A solution would be to add support for quoted strings, or to change the order of the `rule` command to `rule <type> <argument> <window title>` (so for the `tag` type, it will be `rule tag 1 Mozilla Firefox`);
* Rules are added directly to `config_root`. Maybe you would like to have a separate group for all the rules?

Looking forward to your feedback!

Unia commented:
While reusing actions for this makes sense from a developer perspective (code reuse, et cetera) it might be confusing for a user. Semantically, a rule is something different than an action. But, I'll leave it up to you to decide on this. Let me know which route you want to take and I'll get back to work!

Matching on app_id is definitely a good suggestion; I took the window title because I knew that existed. What other properties are there, and how can I read them? Is there something like `xprop` for wayland/swc?

Quoting window titles seems like the best solution. Another one would be to use `strstr` and allow matching on substrings.

michaelforney commented:
I think actions might even be more flexible, and I don't think it would be too confusing.

For example, you could do

    rule app_id st tag.1.apply
    rule app_id st tag.1.active

which would both move the window to tag 1, and then switch to tag 1.

Or you could have gimp windows set the layout to floating or something, or you could run a certain command when a particular application is opened.

Can you think of anything you might want as a rule-type that could not be expressed as an action?

In terms of window properties, there is not something like xprop. You can look at the [xdg-shell protocol](https://github.com/michaelforney/swc/blob/master/protocol/xdg-shell.xml#L171) for more details. I think it is just title and app_id.

I can work on making actions take a parameter this weekend (this might be also useful for key bindings, which might want to know which key triggered them).

Unia commented:
> Can you think of anything you might want as a rule-type that could not be expressed as an action?

Not directly, I was only expressing the semantic difference. I like your solution to reuse actions, though.

I'll update my work once you've implemented parameters for actions :)

michaelforney commented:
I pushed a branch `variant`, which you should be able to use in your rules implementation. Let me know if this works for you.

Unia commented:
I wasn't really sure what you wanted, so at the moment I have two versions for you to see:

* rules branch, this diff. Syntax: `rule st tag tag.2.apply`.  
  This uses the `rule` keyword in the config file, while internally using `handle_action` to parse them. The identifier is the window title or app_id, both of which at the moment are checked with `strstr` (which works and is an easy fix). The `type` is `tag`, which can be extended later with more types. However, the biggest problem with this branch is that there is no way to enforce that a rule with type `tag` doesn't do something other than tagging.
* [rules2 branch](https://github.com/Unia/velox/tree/rules2). Syntax: `action st rule tag.2.apply`.  
  This just adds a new action type, `rule`; solving the issue that we cannot enforce what action the user uses. The identifier is still the window title or app_id, both still checked with `strstr`.

Some things to note for both versions:
* We need to check if `window->swc->app_id` is not `NULL`.
* Using `tag.?.activate` in a rule crashes velox. I have not yet figured out why.
* Perhaps add a config group for rules, to speed up iterating through all of them?

The code in both versions could probably use some cleanup, but I would like to know which version to continue with before I do that and fix the rest. Let me know what you think!

michaelforney commented:
The first branch is closer to what I had in mind. Sorry for the confusion.

I added a comment to your `rules` branch commit. Let me know if that makes sense or not.

Unia commented:
I updated my PR. Is this closer to what you had in mind? (Ignoring possible code cleanups; I know I can improve some areas but wanted to get a general ok first)

Unia commented:
Updated again; did some of the cleanups and fixed an embarrassing oversight.

michaelforney commented:
This looks really good!

I added a first round of comments. Many are nitpicky; sorry about that :)

Unia commented:
Damn, GitHub didn't move your comments over. Find them [here]().

All are taken care of, except for this one:
> You should check if this (config.c, line 424: the `strdup` call) returned NULL.
> Usually the technique I use for things like this is to "unwind" using error labels.
> In this case,
> ```
> error2:
>   free(action->needle);
>error1:
>    free(rule);
>error0:
>    return false;
```

There are two other `strdup` calls in `config.c` where the return value also is not checked. Do I update these, as well?

Unia commented:
^ Fixed, but I'm not really liking the solution: I'm relying on specific free(3) behaviour...

michaelforney commented:
No need to fix the other `strdup` calls. I can do that later today (thanks for spotting them though).

You should add your name to the header of the various files you modified.

Unia commented:
Updated everything to incorporate your comments. I also added a [print](https://github.com/Unia/velox/commit/6229b6ac114d56871c17e3efdb999c4e671f8f7d#diff-e3525ca4a1a338ed0115251c17b21abeR429) in case the type is unknown.

Thanks for your guidance and patience :)

michaelforney commented:
Merged with some very small tweaks.

- I reordered the initialization of `struct rule` to match the order of the fields. This had the benefit of removing the need for one of the error labels. I also added some extra braces (I prefer that if/else statements either all use braces, or none).
- I omitted the name for the `rule_type` enum, since it was not used anywhere.
- I dropped the change from `const char *command` to `char *command` since it was no longer necessary after your rebase. I agree that it should be `char *command` because the strings have been allocated with `malloc`, but that should be fixed separately.
- I left out the change to `velox.conf.sample`, because I figure users might base their initial config off of it, and might be confused if their terminals always open on tag 2. Instead, could you send me another pull-request with a description of the new command in `README.md`?

Thanks for your patience with my nit-picking.

Some final thoughts:

I noticed that all my wayland clients had `NULL` titles and app IDs when they were created. This is because the `xdg_shell` implementation in swc creates the window on the first call to `get_xdg_surface`. I re-read the xdg-shell protocol definition, and I think the intention is that the window shouldn't be "managed" until the first buffer is attached. However, I think there is a slight problem with that from a tiling window manager's perspective. We may want to apply rules to decide whether the window should be floating or tiled (or possibly which monitor it is on), but we need to make this decision and communicate it with the client *before* it attaches its first buffer. There doesn't appear to be a way for the compositor to know when it has set (or chosen not to set) the `app_id` and `title`, except by waiting for the first buffer. I will start a thread on wayland-devel about this.

Also, I think that swc does not currently update the app ID for Xwayland windows based on `WM_CLASS`. This should be pretty easy if you want to tackle it.

michaelforney commented:
Sent http://lists.freedesktop.org/archives/wayland-devel/2015-September/024340.html.

Unia commented:
I have sent you some more pull requests to fix things discussed in this PR. Besides those I just sent, there are still two issues to resolve:

* Quoted window titles or application ids;
* Using `tag.?.activate` still crashes velox when `?` is not the active tag.

I will look into these issues and I'll also take a look at the `WM_CLASS` thing. Those will be separate PRs, as well.
